### ISSUE TITLE
Insufficient rationale

### ISSUE THREAD

**Issue by simon-friedberger** [w3c/webappsec-dbsc#13]
1. The following doesn't go into detail on if these are symmetric or public key operations and if there are significant changes between older and newer TPMs.
    > Studies have also been done on the current populations of TPMs, both for latency and for predictability. Currently the average is around 0.5 seconds for signing operations, but with large variations. There are also significant errors in operations, but well below 1%.

2. The threat model should be clarified. My current understanding is that the attacker is
    1. Getting local access. This should be a high bar.
    2. Not given persistent access. This should not be significantly harder than the first step.
    3. Not allowed to do a lot of other things which seem feasible with local access like
        - Steal log-in data
        - Create sessions which don't use DBSC
 
3. Some data on how often such session stealing happens and and estimate of how often they would be prevented should this proposal be implemented. Combined with the narrow kind of attacks this prevents I currently don't see a good justification for the engineering effort.

4. A more detailed comparison with WebAuthN is missing. It's true that WebAuthN is complicated but it also prevents more authentication based attacks than DBSC.

**Comment by arnar** [comment]
Thanks for the thoughtful review, and apologies it's been a minute.

> The following doesn't go into detail on if these are symmetric or public key operations and if there are significant changes between older and newer TPMs.

We added more accurate numbers in https://github.com/WICG/dbsc/blob/main/README.md#tpm-considerations

> The threat model should be clarified ...

The typical adversary that this proposal addresses is the author/operator of the info stealer malware usually operating as MaaS (malware-as-a-service), eg. [link](https://www.first.org/resources/papers/conf2023/FIRSTCON23-TLPCLEAR-Kim-Info-Stealer-Most-Bang-for-the-Buck-Malware.pdf). The malware distribution is largely indiscriminate and triaging the accounts for value happens post-compromise. I.e. the "local access" here is generally not a targeted exploit, but routinely installed and mass-distributed malware, for which the bar is unfortunately not very high.

By disguising as legitimate software, the malware generally obtains user-level privilege. This is often enough to obtain access to credentials stored on disk from other apps, e.g. on Windows browsers don’t have robust options to protect its storage from such local attack. The burden of verifying the legitimacy of software falls mostly onto the Internet user which is a particularly difficult task if the software itself they’re looking for is from grey area (cracked, unlicensed, adult etc) and provided to them by content platforms (ie. other users with potentially malicious intent). Browsers and antivirus engines can flag many of these at download time, but it is an arms race and attackers use social engineering to convince users to bypass those warnings. 

Regarding persistence and stealing login credentials instead, the goal is that device binding session is part of a bigger picture. It constrains attackers in ways that makes them more detectable, makes active mitigations and cleanup easier, and/or shifts the goal to authentication where the OS and user can afford more time or interaction.

It's correct that persistence isn’t technically harder than one-off access, but it does seem significantly more detectable. We found that malware authors avoid persistence likely in order to avoid detection by agents running on device (and subsequent attribution). Persistent access is less reliable and more visible to network administrators, proxy or hosting operators, making this threat less attractive for more targeted cases.

DBSC itself does not address theft of login credentials. However the ecosystem of login-time protection mechanisms is more mature and also a significantly different problem. Users (or their account administrators) can rely on eg. 2FA or passkeys that make remote or on-device login significantly harder without user interaction. Identity providers can leverage the indiscriminate nature of the malware to detect scaled credential stuffing. Finally, users themselves can better self-diagnose login-centric attacks on their accounts than the "under the hood" session auth. For instance users may be more accustomed to inspecting login events for their identity providers, while session theft is more insidious because it piggybacks on the existing session and it is easier to hide abuse in the noise.

Creating sessions which don’t use DBSC (“unbound” sessions) is a threat we are actively looking into. There are two cases:

 - Stealing login credentials to sign-in on behalf of the victim to an unbound session; we covered it above.
 - Downgrade or MiTM attack around the users’ DBSC session creation (usually login). We believe that the need of engineering user interactions will increase complexity and the detection by users easier. Also the obvious downgrade pattern will make it easier to detect and block by identity provider or account administrator.
 
At the end of the day, exfiltrating sessions off device is a tool for attackers to bypass many of these areas, detection, sign-in methods, malware cleanup, etc. Without device binding, many of the benefits and developments around these are lost. So we believe DBSC or any kind of session binding has significant value, as an important missing piece of the puzzle.

> Some data on how often such session stealing happens and and estimate of how often they would be prevented should this proposal be implemented. 

We assume that all information stealers are capable of stealing session credentials and their distribution numbers go into millions per month, according to infosec reports: https://www.secureworks.com/research/the-growing-threat-from-infostealers, https://www.kelacyber.com/wp-content/uploads/2023/05/KELA_Research_Infostealers_2023_full-report.pdf. At Google we observe attacks on session tokens of the same order of magnitude as password-based attacks. Once installed, malware indiscriminately steals authentication data from any major browser and any major identity provider, so installation numbers provide an upper bound on the scale of threat.

We believe this proposal will have a disruptive impact on session stealing because shorter-lived bearer tokens are harder to commodify and pushing malware authors away from MaaS towards AaaS (Access-as-a-service) will be easier to detect and therefore much harder to scale, as argued above.

> A more detailed comparison with WebAuthN is missing. It's true that WebAuthN is complicated but it also prevents more authentication based attacks than DBSC.

We see the two efforts as solving very distinct problems. WebAuthn is strictly scoped to authentication of users, with credentials that are meant to outlive individual sessions. WebAuthn ceremonies are always interactive with the user, and the properties of how credentials are stored is deliberately left for the user to control - e.g. by storing them on USB security keys, or (virtually) as passkeys in a password manager. WebAuthn does not try to solve the problem of device-binding sessions on an ongoing basis after the user is signed in.

It would be possible to add functionality to WebAuthn to specifically support local device-bound credentials meant for silent (i.e. no UI) authentication suitable for session binding. As you also mentioned, WebAuthn has significant complexity in order to provide the flexibility needed around user authentication credentials, interactive ceremonies, the separation of client and authenticator concepts, attestation, extensions, etc.

Binding sessions to devices is a much simpler problem, and we think it would be a disservice to developers to try and combine them. The complexity in DBSC comes mainly from solving the problem of how to retrofit it to existing cookie-based session implementations, and driving the periodic refreshes without significantly complicating frontend web apps. Neither of these problems exist in WebAuthn, i.e. there is little benefit to WebAuthn's goals to introduce these mechanisms there.

In short, WebAuthn is about what happens before the user is authenticated, and DBSC is about what happens after. WebAuthn's problem involves the user making decisions, but DBSC does not. For those reasons we think the problem spaces are disjoint enough that they are best solved by separate APIs.

That said, we do think there are future integrations between the two, where e.g. a WebAuthn assertion carries information about the device-bound key that drives the subsequent DBSC session, i.e. together the two can help ensure that a session is bound to precisely the device where the user signed in with phishing-resistant WebAuthn credentials. But these will be future addons to our current proposed starting point for DBSC.

**Comment by simon-friedberger** [comment]
Given that you assume user-level access, could this be implemented by an operating system key store to support users without a TPM?

**Comment by arnar** [comment]
Yes, if there are non-TPM based mechanisms that can protect keys better than cookies (that need to be readable by the UA), then that definitely makes sense. For instance we are keenly watching things like https://techcommunity.microsoft.com/t5/windows-it-pro-blog/advancing-key-protection-in-windows-using-vbs/ba-p/4050988

**Comment by simon-friedberger** [comment]
Assuming this can be done with an operating system keystore instead of a TPM, could this solve the performance issues of TPMs in a simpler way?

**Comment by kmonsen** [comment]
Yes, this will work around the performance and reliability issue of TPMs. The main concern is availability, our plan is to support it as soon as it makes sense. We are also open to other storage solutions as long as they offer meaningful better protection that cookies currently have.

I think there is an opportunity to make a solution that makes sense here for each browser and operating system. It does not need to be the same for everyone.

**Comment by NickBorgersOnLowSecurityNode** [comment]
+1 to not relying on TPM for this.

I am concerned that performance limitations associated with TPMs will encourage implementing servers to re-validate less frequently. For big players with large security teams (Okta, Meta, etc.) I bet the server's choices for re-validation will be sensible, but the criteria for re-checking the device binding is up to the implementer so choices will vary. If events occurs where users are negatively impacted by too-frequent or too-many re-validations, implementers will likely shy away from doing it "too often" and potentially degrade the value. 

On my first read of this proposal, change of client IP seems a good example of when the device binding should be re-validated, but that's an event which impacts every session on a device. If it takes a full second of interruption because every one of a user's dozen sessions need to be re-validated, user's won't be happy and servers will adjust their configs. If a dozen re-validations occur in the CPU there should be ample resources in our modern multi-core architectures, if it's running through a TPM I'm concerned.

**Comment by arnar** [comment]
I think we're in agreement here as well. I regret if the explainer makes it seem TPMs are a key part of this long-term. In general we'd very much like to make it so that the API/protocol does not distinguish, or possibly not even allow the website to distinguish, between different key storage/protection mechanisms. These are very different between different platforms, and will vary and improve over time. Personally I believe availability of deployable solutions (which I hope DBSC becomes) also will help drive such improvements.

So in general our view is that it is on the implementor of the client side, in other words the browser, to choose the best tradeoff between how secure a key is against different attacker classes, and how often it can be used. This does add complexity around e.g. the protocol maybe needing to support things like rate-limiting, or refreshes not always happening right away when needed. But the upshot is that browsers can adjust to different clients, e.g. some older clients will never have better key protection than a slow TPM, and we'll continue to have such clients in large numbers for some years. 
Offering DBSC _only_ in cases where the signer is "fast enough" does not seem likely to succeed right now.

There's quite a lot of flexibility there for browsers, e.g. they can choose software-based kernel-level protections where available, while using TPM or similar hardware on other devices. They can also adjust how they implement and schedule the pre-emptive refreshes based on that performance, how many websites have live sessions, their usage patterns, etc.

In my view it's better to let browsers deal with that complexity, and adjust to how the landscape of system-provided key protections might change over time, than to e.g. defer decisions on that to websites.

**Comment by simon-friedberger** [comment]
If we assume there is an OS keystore which is fast enough, isn't this proposal over the top and it would be better to just use signatures?

**Comment by danmarg** [comment]
Most Windows machines _today_ do not have the capabilities for VBS--Statista tells me that Win11 is about 25% of Windows users (and there are [additional](https://learn.microsoft.com/en-us/windows-hardware/design/device-experiences/oem-vbs) hardware requirements for VBS, but they might be trivial?).  (Just to be clear: software-only key storage, e.g. kernel-mode or a dedicated Windows service, is insufficient for defending against modern malware, which typically obtains privileged access.)

Anyway, yes, I hope that in some years most machines have this capability. But what would it change in the protocol?

- I think you'd still need a way to fetch a new server challenge on an ongoing basis to defend against attackers being able to replay or pre-generate signed key proofs (while the attacker has temporary access to the signing device).[1]
- Those server challenges seemingly need to be fresh--i.e., you can't just pre-fetch a bunch and store them to sign later, since they then fail to guarantee recency (i.e they are subject to the same pre-signing attack).
- So now you have the need to periodically refetch some cryptographic material (in this case, the challenge nonce) from the server, probably in a background sort of request, in anticipation of future requests. This sounds basically like the background refresh protocol already!

I think the reason this design still arises is because even if signing is itself fast, fetching the challenge is not. So you *still* have to figure out how to optimize that part of the protocol to avoid user-visible latency.[2]

There are other designs in this space (like sending new challenges across opportunistically, in regular HTTP traffic, and having some magical error condition for the case where a client comes back without having seen a new challenge for a period of time), but they don't strike me as necessarily simpler than the current design and, importantly, do not actually depend upon the signing latency.

Does that make sense?


1. As an aside, you could just say, no, we're going to sign whatever is in the request headers, and if servers don't want replays, they should use CSRF tokens, but that strikes me as a sort of weird layer violation--"DBSC doesn't provide the security you expect unless you also do this application-layer thing."
2. Signing being fast still gives you *other* advantages: you can sign every request, which I think is a nice security property. But it doesn't suggest you can do away with periodically rotating the server-provided challenge.

**Comment by simon-friedberger** [comment]
You can just sign a counter (or another time-like thing) and invalidate all signatures of counter values less than the last one you have seen. Or you include new challenges in response headers. Both give the attacker at most one replayable message which is probably fine in a threat model where the attacker has temporary access anyway.

This eliminates the additional infrastructure and the latency by the additional requests. Allowing implementation in some kind of key store as @arnar suggests is also more portable than requiring a TPM and gives the browser the flexibility to pick something that is fast and secure enough.

Afair at some point the explainer actually said something along lines of "we would like to sign every request but the TPMs are too slow".

So, will this still be a good solution once TPMs are fast enough or VBS-like solutions are available more broadly?

**Comment by danmarg** [comment]
Thanks for that. 

Yes, you can do counters or timestamps.[^1] However, those fundamentally alter the threat model: an attacker can now pregenerate signed responses while having temporary access to the device, and use them in the future off the device.[^2]

You _can_ include new challenges in response headers--that part is true, and it might make sense as an option to opportunistically eliminate refresh requests. 

But, stepping back a bit: as long as you want the specific property of "requests must come with a _recent_ proof of possession of the private key", you have to be able to require authenticated requests to _go get a new challenge from the server_ prior to them being authenticated--AFAICT there's no other way to achieve that specific security property--which means you have to support something that looks vaguely like a "refresh" pattern. (It could be, as I said above, a "here's an error and a challenge; sign it and try again" pattern instead, but that's not fundamentally that different.)

I do still think signing every request is a nice option to have[^3], and to be totally clear, I think it's absolutely ideal if browsers can use lower latency options for key storage than TPMs whenever possible. But I don't (yet) see how to achieve the desired security properties, specifically around a _time bound_ proof of key possession, without some sort of "refresh"-like pattern. 

Do you? 

If I'm right above--and I'm happy to be wrong--then I think the TPM latency is not _really_ the driver of the "refresh" pattern or resultant protocol complexity. 

(As an alternative framing: what would the protocol look like if we didn't have TPM latency as a consideration but _did_ want the same security properties? I _think_ it would look basically the same, but I might be missing some better solution!)



[^1]: As a nitpick, though, I think these can be harder for servers to implement. Whereas a server-provide challenge can be implemented either via a server-side database or in a fully distributed manner--servers can just sign the challenge with a secret key--with the counter approach, servers have to maintain and update globally consistent server-side state. While they don't need _strong_ consistency and can take shortcuts on e.g. only updating the state every few requests--which itself has security tradeoffs--literally incrementing a server-side counter, in a globally consistent way, at, say, Google scale, would be very expensive.

[^2]: And, say, deleting the user's session, so that the server cannot detect desynchronized counters.

[^3]: In particular, I think if you sign every request, you prevent attackers from grabbing a short-lived bearer token and then using it off-device.

**Comment by simon-friedberger** [comment]
I certainly agree in principle, and while this shouldn't be part of any eventual standard, considered alternatives should definitely be elaborated on in the explainer.

The desired security properties could also be phrased as "make it harder to use stolen cookie information". So, the source of the challenge in "this challenge is too old" becomes more nuanced.

Especially, given that TLS 0-RTT and HSTS-preload and the public suffix list or the [asynchronous check for safebrowsing](https://blog.chromium.org/2024/02/optimizing-safe-browsing-checks-in.html) are all efforts to avoid additional requests when connecting to a site and this now requires an additional request to get a challenge.

**Comment by danmarg** [comment]
Yep, totally agreed that we should document this. 

And to be clear, I think we can optimize away some of those extra round trips by opportunistically serving "upcoming challenges" in regular web traffic, as I think I said above (or intended to say!). 

But in general, with the exception of "returning to a website after the bearer token has expired", our extra RTs are all async, so the latency per se shouldn't be a concern. Conversely, where you are in the "token expired" path, the extra RTs are required to obtain the security properties we want, and are unfortunately blocking--and perhaps half-ish of that latency comes from the TPM, which is why we're so eager to use alternatives like VBS instead. 

So just to make my commentary above slightly more nuanced, I do think VBS will be a great improvement, and I do think there's a risk of user-visible latency from DBSC in certain flows; my only point here was that I don't think the protocol is introducing substantially extra complexity that VBS will let us eliminate.

**Comment by drubery** [comment]
Thanks for the feedback! https://github.com/w3c/webappsec-dbsc/pull/130 will now explicitly state that the TPM requirement comes from Chrome's current implementation and not from the protocol. We are indeed interested in other OS-mediate key storage.
